File: 3_算法框架_套路_labuladong/二叉树集锦/105.通过前序和中序遍历结果构造二叉树.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
class TreeNode {
  constructor(val) {
    this.val = val
    this.left = this.right = null
  }
}
const preorder = [3, 9, 20, 15, 7]
const inorder = [9, 3, 15, 20, 7]
// 主函数
var buildTree = function (preorder, inorder) {
  const build = (preStart, preEnd, inStart, inEnd) => {
    // base
    if (preStart > preEnd) return null
  
    // root 节点对应的值就是前序遍历数组的第一个元素
    let rootVal = preorder[preStart]
    // rootVal 在中序遍历数组中的索引
    let index = 0
    for (let i = inStart; i <= inEnd; i++) {
      if (inorder[i] === rootVal) {index = i  
        break }
    }
    // index为root的索引
    let leftSize = index - inStart
  
    // 先构造出当前根节点
    const root = new TreeNode(rootVal)
    // 递归构造左右子树
    root.left = build(preStart + 1,preStart + leftSize, 
                      inStart,index - 1)
  
    root.right = build(preStart + leftSize + 1,preEnd,
                      index + 1,inEnd
    )
    return root
  }
  return build( 0, preorder.length - 1, 0, inorder.length - 1)
}


console.log(buildTree(preorder, inorder))

File: legacy/1.算法框架_labuladong/3.二叉树集锦/105.通过前序和中序遍历结果构造二叉树.java
class Solution {
	private Map<Integer, Integer> inOrderMap;
	int[] preorder;
	int[] inorder;

	public TreeNode buildTree(int[] preorder, int[] inorder) {
		this.preorder = preorder;
		this.inorder = inorder;

		int n = preorder.length;
		inOrderMap = new HashMap<Integer, Integer>();
		for (int i = 0; i < n; i++) {
			inOrderMap.put(inorder[i], i);
		}
		return myBuildTree(0, n - 1, 0, n - 1);
	}

	public TreeNode myBuildTree(int preStart, int preEnd, int inStart, int inEnd) {
		if (preStart > preEnd) {
			return null;
		}
		int rootIndex = preStart;
		int rootVal = preorder[rootIndex];
		TreeNode root = new TreeNode(rootVal);
		int inRoot_Index = inOrderMap.get(rootVal);
		int size_left_subtree = inRoot_Index - inStart;

		root.left = myBuildTree(preStart + 1, preStart + size_left_subtree, inStart, inRoot_Index - 1);
		root.right = myBuildTree(preStart + size_left_subtree + 1, preEnd, inRoot_Index + 1, inEnd);
		return root;
	}
}
/*
 * 
 * 对于任意一颗树而言，前序遍历的形式总是
 * [ 根节点, [左子树的前序遍历结果], [右子树的前序遍历结果] ]
 * 
 * 即根节点总是前序遍历中的第一个节点。而中序遍历的形式总是
 * [ [左子树的中序遍历结果], 根节点, [右子树的中序遍历结果] ]
 * 
 */
